[DEBUGGER] Disassembler copy whole line option

Nitch2024 opened the pull request:
[DEBUGGER] Disassembler copy whole line option in the right click menu copying address, code and memory pointed to in clipboard all in one go making it easier to research and document

Example: 80313ec4 lwz      r0, 0x000C (r29) targetting 80533674

dreamsyntax commented:
@dolphin-emu-bot rebuild

TryTwo commented:
We've made it a habit to use fmt::format for making strings.  Like
```
std::string text  = fmt::format("{:08x} {}.", address (as a u32), text);
and
text += fmt::format(" targeting {:08x}, target_address");
```
Then you can skip all the intermediate QString stuff. and just convert at the end.

TryTwo commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -826,6 +827,35 @@ void CodeViewWidget::OnCopyCode()
+
+  const QString textAddress = QStringLiteral("%1").arg(addr, 8, 16, QLatin1Char('0'));
+
+  const std::string textCode = [this, addr] {
```
This shouldn't be needed, just save it to a string once. and use the string.

Nitch2024 commented:
> We've made it a habit to use fmt::format for making strings. Like
> 
> ```
> std::string text  = fmt::format("{:08x} {}.", address (as a u32), text);
> and
> if (target_addr)
> text += fmt::format(" targeting {:08x}, target_address.value()");
> ```
> 
> Then you can skip all the intermediate QString stuff. and just convert at the end.

Ok, no problem. I will make a quick pass.

Nitch2024 commented:
> We've made it a habit to use fmt::format for making strings. Like

Ok it fits better the style of the codebase. Let me know if you want me to change anything else.

TryTwo commented:
That looks better, the only other thing is just the code snippet I marked above.

Nitch2024 commented:
> That looks better, the only other thing is just the code snippet I marked above.

I am happy to fix it, just not sure I understand.

"> @@ -826,6 +827,35 @@ void CodeViewWidget::OnCopyCode()
   QApplication::clipboard()->setText(QString::fromStdString(text));
 }
 
+void CodeViewWidget::OnCopyWholeLine()
+{
+  const u32 addr = GetContextAddress();
+
+  const QString textAddress = QStringLiteral("%1").arg(addr, 8, 16, QLatin1Char('0'));
+
+  const std::string textCode = [this, addr] {

This shouldn't be needed, just save it to a string once. and use the string."

What part do you recommend I save once?

TryTwo commented:
Sorry, maybe the snippet tool messed up?
```
  const std::string textCode = [this, addr] {
    Core::CPUThreadGuard guard(m_system);
    return m_system.GetPowerPC().GetDebugInterface().Disassemble(&guard, addr);
  }();
```

Can just be  
`` Core::CPUThreadGuard guard(m_system);``
`` const std::string textCode  = m_system.GetPowerPC().GetDebugInterface().Disassemble(&guard, addr);``

or if you want to enclose the guard:
```
std::string textCode
{
   Core::CPUThreadGuard guard(m_system);
   textCode  = m_system.GetPowerPC().GetDebugInterface().Disassemble(&guard, addr);
}
```

Oh and we also have a variable naming convention: so it should be text_code or code_text.

sepalani commented:
> Sorry, maybe the snippet tool messed up?
> 
> ```
>   const std::string textCode = [this, addr] {
>     Core::CPUThreadGuard guard(m_system);
>     return m_system.GetPowerPC().GetDebugInterface().Disassemble(&guard, addr);
>   }();
> ```

^ IIRC, this idiom is used in functions above and allows to mark the variable as `const`. However, if the variable is modified later on, I don't think there is much more benefit to keep it, so your alternatives might be more suited:

> Can just be ` Core::CPUThreadGuard guard(m_system);` ` const std::string textCode = m_system.GetPowerPC().GetDebugInterface().Disassemble(&guard, addr);`
> 
> or if you want to enclose the guard:
> 
> ```
> std::string textCode;
> {
>    Core::CPUThreadGuard guard(m_system);
>    textCode  = m_system.GetPowerPC().GetDebugInterface().Disassemble(&guard, addr);
> }
> ```
> 
> Oh and we also have a variable naming convention: so it should be text_code or code_text.

sepalani commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -826,6 +827,31 @@ void CodeViewWidget::OnCopyCode()
 
+void CodeViewWidget::OnCopyWholeLine()
+{
+  const u32 addr = GetContextAddress();
+
+  const std::string textCode = [this, addr] {
+    Core::CPUThreadGuard guard(m_system);
+    return m_system.GetPowerPC().GetDebugInterface().Disassemble(&guard, addr);
+  }();
+
+  std::string wholeLineText = std::format("{:08x} {}", addr, textCode);
+
+  if (IsInstructionLoadStore(textCode))
+  {
+    const std::optional<u32> target_addr =
+        m_system.GetPowerPC().GetDebugInterface().GetMemoryAddressFromInstruction(textCode);
+
+    if (target_addr)
+    {
+      wholeLineText += std::format(" targetting {:08x}", *target_addr);
+    }
+  }
+
+  QApplication::clipboard()->setText(QString::fromStdString(wholeLineText));
```
The following should fix your issue as we're using fmt's format instead of std.
```suggestion
  const u32 addr = GetContextAddress();

  const std::string text_code = [this, addr] {
    Core::CPUThreadGuard guard(m_system);
    return m_system.GetPowerPC().GetDebugInterface().Disassemble(&guard, addr);
  }();

  std::string whole_line = std::format("{:08x} {}", addr, text_code);

  if (IsInstructionLoadStore(text_code))
  {
    const std::optional<u32> target_addr =
        m_system.GetPowerPC().GetDebugInterface().GetMemoryAddressFromInstruction(text_code);

    if (target_addr)
    {
      whole_line += fmt::format(" targetting {:08x}", *target_addr);
    }
  }

  QApplication::clipboard()->setText(QString::fromStdString(whole_line));
```

TryTwo commented:
> ^ IIRC, this idiom is used in functions above and allows to mark the variable as `const`. However, if the variable is modified later on, I don't think there is much more benefit to keep it, so your alternatives might be more suited:

Oh, I didn't see that we used it elsewhere. I thought it weird the function/guard was being hit 3 times in a row. Unless it's optimized into one hit? Either way, since it runs once on user interaction it probably doesn't matter much.

Nitch2024 commented:
> > ^ IIRC, this idiom is used in functions above and allows to mark the variable as `const`. However, if the variable is modified later on, I don't think there is much more benefit to keep it, so your alternatives might be more suited:
> 
> Oh, I didn't see that we used it elsewhere. I thought it weird the function/guard was being hit 3 times in a row. Unless it's optimized into one hit? Either way, since it runs once on user interaction it probably doesn't matter much.

Yes I used the same construct that was everywhere else assuming there must have been a reason for it to exist, but I simplified since the local scope of the function seems fine for the life time of the guard and given that this is done when the game is paused in the code view.

Addressed all feedback:

- Using fmt::format
- Changed variable names to match coding convention
- Changed the cpu guard code

Let me know if I missed anything

TryTwo commented:
LGTM

TryTwo commented:
Actually, some lines come out with a tab in the middle ``804d0c7c rlwinm	r3, r3, 17, 31, 31 (00008000)``. You can do something like ``std::replace(string.begin(), string.end(), '\t', ' ');``  Not sure if it needs to be replaced with a space or empty.

Also you can close suggested changes that are fulfilled.

Nitch2024 commented:
Better not to remove it from the clipboard, if you copy many lines the tabs help retaining alignment for visibility wherever you paste it. I removed the tab in the assemble dialog for space issues, but here it seems like a better choice to keep all the tabs.

Nitch2024 commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -826,6 +827,29 @@ void CodeViewWidget::OnCopyCode()
+
+  Core::CPUThreadGuard guard(m_system);
+
+  std::string text_code = m_system.GetPowerPC().GetDebugInterface().Disassemble(&guard, addr);
```
Yes, good catch. Fixing it now
